### PR TITLE
add option to preserve servers in server.xml file not found in the provi...

### PR DESCRIPTION
**Pull request by NATHAP**
...ded list of credentials

I have a use case where I really need to preserve server entries in the servers.xml file which I can't put into the credentials list, so here's some code.

### CODE REVIEW

**Comment by imod** [comment]
do we really need the option to replace all? would it not make sense to not touch the ones not found in the credentials?

**Comment by NATHAP** [comment]
I agree with you about it making sense not to touch the ones not found in the credentials, but didn’t want to break existing behaviour.
Actually at first I thought it was a bug, but then realised it was expected behaviour – the plugin page says…

“…all server configs in the same file will be removed and replaced…”

In the end I decided to…


1.       add a ‘replace all’ option

2.       default the ‘replace all’ to maintain existing behaviour

3.       describe the option as ‘replace all’ as it seemed like a fairly simple and intuitive phrase for the ui

File: src/main/java/org/jenkinsci/plugins/configfiles/maven/security/CredentialsHelper.java [hunk 1]
```diff
@@ -150,17 +150,32 @@ public static String fillAuthentication(String settingsContent, Map<String, Stan
...
-        for (int i = 0; i < childNodes.getLength(); i++) {
-            final Node child = childNodes.item(i);
-            serverNode.removeChild(child);
-            --i;
+    private static void removeAllChilds(final Node serversNode, final Set<String> credentialKeys, final Boolean replaceAll) {
+        final NodeList serverNodes = serversNode.getChildNodes();
+        for (int i = 0; i < serverNodes.getLength(); i++) {
+            final Node server = serverNodes.item(i);
+            String serverId = getServerId(server);
+            if (Boolean.TRUE.equals(replaceAll) || (credentialKeys.contains(serverId) && Boolean.FALSE.equals(replaceAll))) {
```
Comment (imod) [inline src/main/java/org/jenkinsci/plugins/configfiles/maven/security/CredentialsHelper.java]: is `&& Boolean.FALSE.equals(replaceAll)` really required? if `serverId` is containedin the key, then we remove it anyway

Comment (NATHAP) [inline src/main/java/org/jenkinsci/plugins/configfiles/maven/security/CredentialsHelper.java]: Good point, thanks for that simplification.

**Pull request merged**